### PR TITLE
add LSP capability textDocument/documentHighlight

### PR DESCRIPTION
**Pull request by lgo**
cc @jvilk-stripe. Changes from #1623 and after rebasing. The text below is largely from the previous PR.

This adds the LSP functionality for `textDocument/documentHighlight` ([Microsoft specification for documentHighlight](https://microsoft.github.io/language-server-protocol/specification#textDocument_documentHighlight)).

When a user selects a piece of text, the LSP responds with a list of ranges to highlight in the current file which relate to the initiate selection. 

Currently, highlighting uses the same information `textDocument/references` does. That means  this will work for typed files otherwise it can only handle Ruby constants in untyped files. Compared to regular fuzzy-highlighting, this will only respond with highlights for an identifier reference within the scope it is valid in which is the added benefit of this LSP method!

This does not implement the response field `kind?: DocumentHighlightKind`, which identifies the response values as the more specific highlights:
- `Text`, a textual occurrence -- typically a fuzzy response.
- `Read`, a read-access of a symbol, like reading a variable.
- `Write`, a write-access of a symbol, like writing to a variable.

By default, the editor will consider no `kind` as a `Text` response.

After poking in VSCode I noticed there are [distinguishing styling for the types](https://github.com/microsoft/vscode/blob/466e5ef6c2188d61a8c93c53748aa9e9788bcfc9/src/vs/editor/contrib/wordHighlighter/wordHighlighter.ts#L427-L452). In particular, the styling settings, [found here](https://code.visualstudio.com/api/references/theme-color#editor-colors), are:
- `editor.wordHighlight{Background,Border}` for read-access
- `editor.wordHighlightStrong{Background,Border}` for write-access
- `editor.selectionHighlight{Background,Border}` for fuzzy text.

Currently, this LSP method has an issue where the range of the highlights is rather generous. The quirks are
- highlighting entire method definition lines (`def` + args)
- highlighting entire method call line (receiver, inputs, input block), eg: `d.method(1, 2) {}`. the same happens for `send` assignments assignments, eg: `d.method = 'bar'`)
- highlighting the `:` when highlighting parameter identifiers, eg: `foo(b:)`

I did some digging into the Locations used around here. As I understand it, we use the declLoc from these that come from the parsing step. Specifically, the problem cases from before are that we just want location of the SymbolRef for these -- and it primarily happened on Declaration nodes and Arg-type Reference nodes. I imagine we'd need to carry through a symbolLoc or know how to lazily compute it.

Is that a correct understanding? I'm just digging to understand what needs to be done to make this LSP feature usable. As you mentioned, let's get this in w/o that fix yet, but I don't think this is generally useful until it's fixed.

### Motivation

LSP features :)

### Test plan

Updated the tests so that all of the `def` + `usage` assertions will also check the highlight query. It differs by only checking the query returns values for all assertions in one file rather than all of them.

### CODE REVIEW

File: test/helpers/position_assertions.h [hunk 1]
```diff
@@ -178,6 +181,14 @@ class TypeAssertion final : public RangeAssertion {
     std::string toString() const override;
 };
 
+// does assertions using positions specified by # ^^^ usage: symbol
```
Comment (jvilk-stripe) [inline test/helpers/position_assertions.h]: It's confusing to have a class named `HighlightAssertion` when there's no `highlight:` assertion.

Could we add it to `UsageAssertion` instead (maybe `checkHighlight`)?

**Comment by jvilk-stripe** [comment]
Oh yeah, and you broke the dumb LSP recorded tests (which we are slowly removing as we improve coverage with other, better tests) -- you'll need to update them.

**Comment by lgo** [comment]
ptal @jvilk-stripe -- addressed comments and fixed up the tests. Or, is it okay to hit merge? Not sure what the norm is in sorbet :)

**Comment by jvilk-stripe** [comment]
@lgo OK to merge!

**Pull request merged**